### PR TITLE
fix: allow to push single items in arrays

### PR DESCRIPTION
**Pull request by Yarmeli**
Implemented a fix for #181

### CODE REVIEW

**Comment by arthurfiorette** [comment]
Can you add some tests?

**Comment by Yarmeli** [comment]
I added some tests, let me know your thoughts on them 😄 

Also, I noticed that some tests that should fail, don't really fail, for example:

```ts
// literal.test-d.ts

expectAssignable<Model>({
  id: 0,
  simple: 1,
  optional: 2,
  list: [3]
}); // ✅ it passes correctly

// copy the above code and replace the method with 'expectNotAssignable'
expectNotAssignable<Model>({
  id: 0,
  simple: 1,
  optional: 2,
  list: [3]
}); // ❌ the test passes but it should fail because this can be assigned to Model
```

My understanding is that this happens because of how arrays are handled in the `tsd` package. I [opened an issue in there](https://github.com/tsdjs/tsd/issues/199)

**Comment by arthurfiorette** [comment]
Should we wait for this fix at tsd?

**Comment by Yarmeli** [comment]
I don't think it would make much of a difference waiting for the tsd fix/suggestion on how to handle arrays since that currently affects new and old tests.

I can open a new issue to improve the tests depending on what the fix is for that issue

**Comment by arthurfiorette** [comment]
@Yarmeli ok. CI is still failing

**Review by arthurfiorette — APPROVED** [review]

File: test/types/mongo.test-d.ts [hunk 1]
```diff
@@ -38,6 +38,34 @@ expectAssignable<Model>({
   list: [3, 3, 3]
 });
 
+expectAssignable<UpdateManyInput<Model["list"][0]>>({
```
Comment (arthurfiorette) [inline test/types/mongo.test-d.ts:41]:
```suggestion
expectAssignable<UpdateManyInput<Model["list"][number]>>({
```
Please change on other places too

File: assets/namespace.d.ts [hunk 2]
```diff
@@ -175,7 +175,7 @@ export type TypedStringNullableListFilter<S extends string> =
  */
 export type UpdateStringArrayInput<S extends string> = {
   set?: S[];
-  push?: S[];
+  push?: S[] | S;
```
Comment (arthurfiorette) [inline assets/namespace.d.ts:178]:
```suggestion
  push?: S | S[];
```

consistency :)

File: assets/namespace.d.ts [hunk 3]
```diff
@@ -27,7 +27,7 @@ export type UpdateInput<T> = T extends object
   : T;
 
 /** A type to determine how to update a json[] field */
-export type UpdateManyInput<T> = T | T[] | { set?: T[]; push?: T[] };
+export type UpdateManyInput<T> = T | T[] | { set?: T[]; push?: T[] | T };
```
Comment (arthurfiorette) [inline assets/namespace.d.ts:30]:
```suggestion
export type UpdateManyInput<T> = T | T[] | { set?: T[]; push?: T | T[] };
```

consistency :)

**Review by arthurfiorette — CHANGES_REQUESTED** [review]

File: test/types/mongo.test-d.ts (hunk 1 above)
Comment (Yarmeli) [reply test/types/mongo.test-d.ts]: Done

**Comment by arthurfiorette** [comment]
Can you run `pnpm prettier --write .` and commit the changes

**Comment by Yarmeli** [comment]
Hopefully, this should have worked 🤞

I am having issues on my local machine with prettier not being able to find prettier plugin modules for this repo 😞

**Review by arthurfiorette — APPROVED** [review]

**Comment by mrazauskas** [comment]
@Yarmeli Just a quick note. Might be you saw already, last week I released TSTyche, a fresh type testing tool. This is the one I promised in https://github.com/tsdjs/tsd/issues/199

Repo: https://github.com/tstyche/tstyche
Documentation: https://tstyche.org

---

Before we talked about widening of inferred types. I explained it in [documentation](https://tstyche.org/reference/expect-api#type-widening). This is how TypeScript works and that is for good reason. But with TSTyche you can avoid widening in your type tests by comparing two types:

```ts
import { expect } from "tstyche";

interface Options {
  locale?: Array<"en" | "de">;
  root?: string;
}

expect<Options>().type.toBeAssignable<{
  locale: ["en", "de"];
}>();

expect<Options>().type.toBeAssignable<{
  root: string;
}>();
```

Also with TSTyche you can test types on specific versions of TypeScript: `tstyche --target 4.8,5.0,latest`. Very useful for complex generic types. Give it a try (;

**Pull request merged**